src/ZkFold/Symbolic/Data/ByteString.hs

```diff
@@ -368,119 +275,77 @@ instance (Arithmetic a, KnownNat n) => ShiftBits (ByteString n (ArithmeticCircui
 
-instance
-  ( Mod k m ~ 0
-  , (Div k m) * m ~ k
-  , Arithmetic a
-  ) => Concat (ByteString m (ArithmeticCircuit a)) (ByteString k (ArithmeticCircuit a)) where
-
-    concat bs = ByteString $ bsCircuit {acOutput = bsOutputs}
+    isUnset (ByteString bits) ix = Bool $ fromCircuitF bits solve
         where
-            bsCircuit = Haskell.mconcat $ (\(ByteString bits) -> bits {acOutput = U1}) <$> bs
+            solve :: forall i m . MonadCircuit i (BaseField c) m => Vector n i -> m (Par1 i)
+            solve v = do
+                let vs = V.fromVector v
+                    i = (!! ix) vs
+                j <- newAssigned $ \p -> one - p i
+                return $ Par1 j
 
-            bsOutputs :: Vector k Natural
-            bsOutputs = V.unsafeConcat @(Div k m) $ (\(ByteString bits) -> acOutput bits) <$> bs
+--------------------------------------------------------------------------------
 
-instance
-  ( KnownNat n
-  , n <= m
-  ) => Truncate (ByteString m (ArithmeticCircuit a)) (ByteString n (ArithmeticCircuit a)) where
+toBsBits :: Natural -> Natural -> [Natural]
+toBsBits num n = reverse bits
+    where
+        base = 2
 
-    truncate (ByteString bits) = ByteString $ bits { acOutput = V.take @n (acOutput bits) }
+        availableBits = unfoldr (toBase base) (num `Haskell.mod` (2 Haskell.^ n)) <> Haskell.repeat 0
 
-instance
-  ( KnownNat m
-  , KnownNat n
-  , m <= n
-  , Arithmetic a
-  ) => Extend (ByteString m (ArithmeticCircuit a)) (ByteString n (ArithmeticCircuit a)) where
+        bits = take (Haskell.fromIntegral n) availableBits
 
-    extend (ByteString oldBits) = ByteString $ circuitF (Vector <$> solve)
-      where
-        solve :: forall i m'. MonadBlueprint i a m' => m' [i]
-        solve = do
-            bits <- runCircuit oldBits
-            zeros <- replicateM diff $ newAssigned (Haskell.const zero)
-            pure $ zeros <> V.fromVector bits
+-- | Convert a number into @base@-ary system.
 
-        diff :: Haskell.Int
-        diff = Haskell.fromIntegral $ getNatural @n Haskell.- getNatural @m
+toBase :: Natural -> Natural -> Maybe (Natural, Natural)
+toBase _ 0    = Nothing
+toBase base b = let (d, m) = b `divMod` base in Just (m, d)
 
 
-instance Arithmetic a => BitState ByteString n (ArithmeticCircuit a) where
-    isSet (ByteString v) ix = Bool $ circuit solve
-        where
-            solve :: forall i m . MonadBlueprint i a m => m i
-            solve = (!! ix) . V.fromVector <$> runCircuit v
-    isUnset (ByteString v) ix = Bool $ circuit solve
-        where
-            solve :: forall i m . MonadBlueprint i a m => m i
-            solve = do
-                i <- (!! ix) . V.fromVector <$> runCircuit v
-                newAssigned $ \p -> one - p i
+-- | A generic bitwise operation on two ByteStrings.
+-- TODO: Shall we expose it to users? Can they do something malicious having such function? AFAIK there are checks that constrain each bit to 0 or 1.
+--
+-- bitwiseOperation
```

Comment:
This can be generalized to `Symbolic c` and used, e.g., in the `BoolType` instance.